test/MultisigTask.t.sol

```diff
@@ -0,0 +1,249 @@
+
+        /// set addresses in MultisigTask contract to a deployed addresses
+        /// contract so that these calls work
+        vm.store(address(task), ADDRESSES_SLOT, bytes32(uint256(uint160(address(addresses)))));
+        MockMultisigTask(address(task)).addAction(
+            addresses.getAddress("ProxyAdmin", getChain("optimism").chainId),
+            abi.encodeWithSignature(
+                "upgrade(address,address)",
+                addresses.getAddress("L1ERC721BridgeProxy", getChain("optimism").chainId),
+                MockMultisigTask(address(task)).newImplementation()
+            ),
+            0,
+            ""
+        );
+
+        vm.mockCall(
+            multisig,
+            abi.encodeWithSelector(
+                IGnosisSafe.getTransactionHash.selector,
+                MULTICALL3_ADDRESS,
+                0,
+                task.getCalldata(),
+                Enum.Operation.DelegateCall,
+                0,
+                0,
+                0,
+                address(0),
+                payable(address(0)),
+                task.nonce()
+            ),
+            /// return a hash that cannot possibly be what is returned by the GnosisSafe
+            abi.encode(bytes32(uint256(100)))
+        );
+
+        vm.expectRevert("MultisigTask: hash mismatch");
+        task.simulate();
+    }
+
+    function testBuildFailsRevertPreviousSnapshotFails() public {
+        address multisig = addresses.getAddress("ProxyAdminOwner", getChain("optimism").chainId);
+        /// set multisig variable in MultisigTask to the actual multisig address
+        /// so that the simulate function does not revert and can run and create
+        /// calldata by calling the multisig functions
+        vm.store(address(task), MULTISIG_SLOT, bytes32(uint256(uint160(multisig))));
+
+        /// set addresses in MultisigTask contract to a deployed addresses
+        /// contract so that these calls work
+        vm.store(address(task), ADDRESSES_SLOT, bytes32(uint256(uint160(address(addresses)))));
+
+        MockTarget target = new MockTarget();
+        target.setTask(address(task));
+
+        /// set mock target contract in the task contract as there is no setter method,
+        /// and we need to set the target contract to a deployed contract so that the
+        /// build function will make this call, which will make the MultisigTask contract
+        /// try to revert to a previous snapshot that does not exist. It does this by
+        /// calling vm.store(task, _startSnapshot SLOT, some large number that isn't a valid snapshot id)
+        vm.store(address(task), MOCK_TARGET_SLOT, bytes32(uint256(uint160(address(target)))));
+
+        vm.expectRevert("failed to revert back to snapshot, unsafe state to run task");
+        task.build();
+    }
+
+    function testRunFailsDuplicateAction() public {
+        /// add duplicate action that will cause a revert
+        MockMultisigTask(address(task)).addAction(
```

Comment:
It wasn't immediately clear to me that this action was already added in a previous test so therefore it'll be a duplicate. 

Refactoring this into a function that's used between 2 tests will reduce repeating code.

Reply:
Building the calldata is used in 3 places total. So that can be a separate function too. Something like: 

```solidity
    function addUpgradeAction() internal {
        MockMultisigTask(address(task)).addAction(
            addresses.getAddress("ProxyAdmin", getChain("optimism").chainId), encodeUpgradeAction(), 0, ""
        );
    }

    function encodeUpgradeAction() internal returns (bytes memory) {
        return abi.encodeWithSignature(
            "upgrade(address,address)",
            addresses.getAddress("L1ERC721BridgeProxy", getChain("optimism").chainId),
            MockMultisigTask(address(task)).newImplementation()
        );
    }
```